Reduce overestimation of minimum UTXO values for Shelley-era addresses.

Anviking opened the pull request:
### Issue Number

ADP-2039

### Summary

This PR uses information about the boundaries of address lengths to reduce our overestimation of minimum UTxO values for Shelley-era addresses.

- [x] Use `BoundedAddressLength key` class to have address length overestimations for Shelley, Byron, Icarus, Shared **individually** for change generation. This reduces the overestimation of minUTxOValue for Shelley change outputs by 29 bytes (86-57).
- [x] When validating user-specified outputs, or replacing `Coin 0` values with the minimum allowed, we now use the **specific address** for the `minimumUTxOValue` calculation, instead of using `maxLengthAddress`. 

### Comments

- We are still using `maxLengthAddress` when calculating the minUTxOValue for outputs in migration. It would be better to use the longest address from the list we are given by the user.

### Outstanding Tasks

- [x] Verify the `BoundedAddressLength` implementation for `SharedKey`. (Recently added.)
- [x] Consider adding `minLengthAddressFor` to `BoundedAddressLength`, and redefining the global `minLengthAddress` constant in a similar way to how `maxLengthAddress` is defined (by taking the minimum of all values across all known key types).
- [ ] Add golden tests for the lengths (in bytes) of addresses returned by `maxLengthAddressFor` and `minLengthAddressFor`. These will fail fast if anything in our underlying implementation changes.
- [ ] Consider adding individual `maxLengthAddressFor` key addresses to the `Address` generators for `MinimumUTxOSpec`.

jonathanknowles committed:
Reduce overestimation of minimum UTxO values for Shelley-era addresses.

Co-Authored-By: Johannes Lund <johannes.lund@iohk.io>

jonathanknowles committed:
Move `dummyAddress` from `SelectionContext` to `SelectionConstraints`.

This allows `dummyAddress` to be specified by the caller of coin
selection instead of it being a hard-coded constant.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1134,7 +1136,15 @@ toApiNetworkParameters (NetworkParameters gp sp pp) txConstraints toEpochInfo =
+            -- address, 9-byte coin, and no tokens.
+            txOutputMinimumAdaQuantity txConstraints
+                (AD.maxLengthAddress $ Proxy @ShelleyKey)
+                TokenMap.empty
```
Unfortunate. Would like to see `Cardano.ProtocolParameters` used for the JSON instead ([related](https://input-output.atlassian.net/browse/ADP-2046?atlOrigin=eyJpIjoiZDc3NTVlZWZkNjllNGQwMzhmZDIwYTQ1NmZhYjk4MTkiLCJwIjoiamlyYS1zbGFjay1pbnQifQ)). Could also use `longestAddressOfAnyType`, but I think some integration test broke then.

jonathanknowles replied:
> Unfortunate. Would like to see `Cardano.ProtocolParameters` used for the JSON instead ([related](https://input-output.atlassian.net/browse/ADP-2046?atlOrigin=eyJpIjoiZDc3NTVlZWZkNjllNGQwMzhmZDIwYTQ1NmZhYjk4MTkiLCJwIjoiamlyYS1zbGFjay1pbnQifQ)). Could also use `longestAddressOfAnyType`, but I think some integration test broke then.

I agree with you, this is very unfortunate. I also agree that we should deprecate this field and use `Cardano.ProtocolParameters` instead.

In the meantime, I've written an explanation for users in this commit: 3a5f5cf0974b52126bfa0c333c4c93006267708a

Anviking commented on lib/core/src/Cardano/Wallet/CoinSelection/Internal/Balance.hs:
```diff
@@ -849,8 +851,12 @@ performSelectionEmpty performSelectionFn constraints params =
 
+
+    -- FIXME: How does the choice of address matter here?
+    dummyAddress = longestChangeAddress constraints
```
If only a dummy address matters here, I think it should be named `Balance.SelectionConstraints.dummyAddress` instead of `Balance.SelectionConstraints.longestChangeAddress`, but we could still pass in a `longestChangeAddress`

Anviking replied:
But it seems we do need `longestChangeAddress` here, because we use it in
```haskell
            { minCoinFor = computeMinimumAdaQuantity longestChangeAddress
```

Anviking replied:
Dropped the `FIXME`.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Migration/Selection.hs:
```diff
@@ -274,7 +276,14 @@ assignMinimumAdaQuantity :: TxConstraints -> TokenMap -> TokenBundle
     TokenBundle c m
   where
-    c = txOutputMinimumAdaQuantity constraints m
+    -- We cannot use 'view #longestChangeAddress constraints' here, as the
```
```suggestion
    -- Using @longestAddressFor $ Proxy @k@ via @constraints@ would not help here, as the
```

Anviking committed:
Update `balanceTx` delegation golden test values.

It's unclear why the error message suggests we need 2508 more lovelace
to generate change. By reducing the minUTxOValue overestimation we
should require /less/ ada for change.

It might be that the 'dummyAddress' used to be involved, and
artificially lowered the minUTxOValue reported in the error message.

It would be nice to eventually test that the lovelace quantities
reported by error messages actually are correct and useful in some
sense: e.g.  increasing the ada quantity by of an input by `shortfall`
should presumably make the operation succeed, and be the lowest amount
which does so.

jonathanknowles committed:
Move `ProtocolMagic` type and functions to separate module.

This is necessary to avoid a circular import.

jonathanknowles committed:
Miscellaneous formatting fixes.

jonathanknowles committed:
Add `BoundedAddressLength` class and instances.

Co-Authored-By: Johannes Lund <johannes.lund@iohk.io>

jonathanknowles committed:
Add `BoundedAddressLength` constraint to functions in `Wallet`.

jonathanknowles committed:
Revise documentation for `minimum_utxo_value` API field.

This commit provides a more detailed explanation of the minimum UTxO
calculation, and how it has changed over the course of several eras.

It also provides guidance on how to use the value reported by this
field.

jonathanknowles committed:
Replace `genCoinRange` with `chooseCoin` in `Migration.SelectionSpec`.

When `SelectionSpec` was written, `chooseCoin` was not yet defined.

But now we have a standard definition for `chooseCoin`, we can use that
to replace function `genCoinRange` within `SelectionSpec`.

jonathanknowles committed:
Rename `dummyAddress` to `maximumLengthChangeAddress`.

This renaming is intended to more accurately reflect the intended usage.

A future commit will ensure that this field is initialized to an
appropriate value.

jonathanknowles committed:
Use a minimum-length address for the dummy output in module `Balance`.

We do need to use a valid address here, since the minimum ada quantity
function requires it, but since the dummy output will eventually be
thrown away, we should use the shortest possible valid address in order
to minimize the (marginal) cost.

jonathanknowles committed:
Fix `STAKE_POOLS_{JOIN,QUIT}_01x` integration tests.
